kimchi/src/circuits/polynomials/foreign_field_mul/gadget.rs

```diff
@@ -0,0 +1,373 @@
+//! This module obtains the gates of a foreign field addition circuit.
+
+use std::collections::HashMap;
+
+use ark_ff::{PrimeField, SquareRootField, Zero};
+use ark_poly::{
+    univariate::DensePolynomial, EvaluationDomain, Evaluations, Radix2EvaluationDomain as D,
+};
+use rand::{prelude::StdRng, SeedableRng};
+use std::array;
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentType},
+        constraints::ConstraintSystem,
+        expr::{self, l0_1, Environment, LookupEnvironment, E},
+        gate::{CircuitGate, CircuitGateError, CircuitGateResult, GateType},
+        lookup::{
+            self,
+            lookups::{LookupInfo, LookupsUsed},
+            tables::{GateLookupTable, LookupTable},
+        },
+        polynomial::COLUMNS,
+        wires::Wire,
+    },
+    curve::KimchiCurve,
+    prover_index::ProverIndex,
+};
+
+use super::circuitgates::ForeignFieldMul;
+
+/// Number of gates in this gadget
+pub const GATE_COUNT: usize = 1;
+
+impl<F: PrimeField + SquareRootField> CircuitGate<F> {
+    /// Create foreign field multiplication gate
+    ///     Inputs the starting row
+    ///     Outputs tuple (next_row, circuit_gates) where
+    ///       next_row      - next row after this gate
+    ///       circuit_gates - vector of circuit gates comprising this gate
+    pub fn create_foreign_field_mul(start_row: usize) -> (usize, Vec<Self>) {
+        let circuit_gates = vec![
+            CircuitGate {
+                typ: GateType::ForeignFieldMul,
+                wires: Wire::for_row(start_row),
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::Zero,
+                wires: Wire::for_row(start_row + 1),
+                coeffs: vec![],
+            },
+        ];
+
+        (start_row + circuit_gates.len(), circuit_gates)
+    }
+
+    /// Create foreign field multiplication gate by extending the existing gates
+    pub fn extend_foreign_field_mul(gates: &mut Vec<Self>, curr_row: &mut usize) {
+        let (next_row, circuit_gates) = Self::create_foreign_field_mul(*curr_row);
+        *curr_row = next_row;
+        gates.extend_from_slice(&circuit_gates);
+    }
+
+    pub fn verify_foreign_field_mul<G: KimchiCurve<ScalarField = F>>(
```

Comment:
this is gone already after your update, right?

Reply:
correct